### PR TITLE
BFD-2299: Replace 500 errors with 400 errors where appropriate and update error messaging

### PR DESCRIPTION
**Pull request by lsmitchell**
**JIRA Ticket:**
[BFD-2299](https://jira.cms.gov/browse/BFD-2299)

**User Story or Bug Summary:**

As a BFD User, I would like the application to give back more useful http status codes and errors in the event that I have given an invalid input to one of the BFD endpoints.

---

### What Does This PR Do?

Changes a number of spots where we would previously throw an `IllegalArgumentException` that would return a 500 into `InvalidRequestException` instead, which will throw a 400 error. Also corrects a number of error messages to be more user friendly and adds a bunch of test coverage for these scenarios.

- Replace `IllegalArgumentException` with `InvalidRequestException` in scenarios where the user input has caused the issue and can be corrected
- Improved some messages to be more helpful with what the issue is and/or how to correct it when the issue is the user's input values/format/etc
- Minor adjustments to make sure error messages for the same scenarios are the same for v1/v2 for consistency
- Misc javadoc fixes on touched code
- Added a few error catches that were not checked/caught into friendly messages in certain places (checking patientId is a valid long, to catch the NumberFormatException that would throw, for example)
- Added some *unit* tests to test the various touched resource providers; these will run much faster than the IT tests. The IT tests are unnecessary for this basic validation checking, and these classes can be a platform for testing things that do not need the entire FHIR server startup to test them, like internal validation.
- Added a file that enables a mockito extension which allows for mocking final classes. This is used in the unit tests to ease the testing setup a bit

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [x] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* This should not be merged before we have partner sign-off and they are ready to recieve these changes; aiming to get this approval by end of January.

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] The data dictionary has been updated with any field mapping changes, if any were made.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

**Comment by lsmitchell** [comment]
Ready for review, but waiting on merge until we verify required partner code is deployed.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/PatientLinkBuilder.java [hunk 1]
```diff
@@ -66,7 +66,7 @@ public PatientLinkBuilder(PatientLinkBuilder prev, boolean hasAnotherPage) {
    */
   private void validate() {
     if (getPageSize() <= 0) {
-      throw new InvalidRequestException("A zero or negative page size is unsupported");
+      throw new InvalidRequestException("Value for pageSize cannot be negative: %s");
```
Comment (mjburling) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/PatientLinkBuilder.java:69]: Zeroes also appear to be invalid here.

**Review by mjburling — CHANGES_REQUESTED** [review]

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/PatientLinkBuilder.java (hunk 1 above)
Comment (lsmitchell) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/PatientLinkBuilder.java:69]: Good catch; not sure why the patient paging doesnt allow 0 and every other page size does, but I'll just adjust the message here to catch that and not second guess the implementation.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4CoverageResourceProvider.java [hunk 2]
```diff
@@ -136,7 +137,10 @@ public Coverage read(@IdParam IdType coverageId) {
 
     Matcher coverageIdMatcher = COVERAGE_ID_PATTERN.matcher(coverageIdText);
     if (!coverageIdMatcher.matches()) {
-      throw new IllegalArgumentException("Unsupported ID pattern: " + coverageIdText);
+      throw new InvalidRequestException(
+          "Coverage ID pattern: '"
+              + coverageIdText
+              + "' does not match expected pattern: {alphaNumericString}-{singleCharacter}-{idNumber}}");
```
[the remark this answers is not in the document]
Comment (lsmitchell) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4CoverageResourceProvider.java]: Yup extra bracket, removed it!

**Review by mjburling — APPROVED** [review]

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/PatientLinkBuilder.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (lsmitchell) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/PatientLinkBuilder.java]: Fixed this in my previous commit.

**Review by mackec1 — APPROVED** [review]

**Pull request merged**